pkg/generate/crds/gen.go

```diff
@@ -0,0 +1,179 @@
+
+	// Search for PROJECT if no rootPath provided
+	if len(c.RootPath) == 0 {
+		c.RootPath, err = filepath.Abs(".")
+		if err != nil {
+			return err
+		}
+
+		for {
+			if pathHasProjectFile(c.RootPath) {
+				break
+			}
+
+			if isGoPath(c.RootPath) {
+				return fmt.Errorf("failed to find working directory that contains %s", "PROJECT")
+			}
+
+			c.RootPath = path.Dir(c.RootPath)
+		}
+	}
+
+	// Switch working directory to root path.
+	if err := os.Chdir(c.RootPath); err != nil {
+		return fmt.Errorf("failed switching working dir: %v", err)
+	}
+
+	if err := b.AddDirRecursive("./pkg/apis"); err != nil {
+		return fmt.Errorf("failed making a parser: %v", err)
+	}
+	ctx, err := parse.NewContext(b)
+	if err != nil {
+		return fmt.Errorf("failed making a context: %v", err)
+	}
+
+	arguments.CustomArgs = &parse.ParseOptions{SkipMapValidation: c.SkipMapValidation}
+
+	// TODO: find an elegant way to fulfill the domain.
+	p := parse.NewAPIs(ctx, arguments, getDomainFromProject())
+
+	crds := c.getCrds(p)
+
+	// Ensure output dir exists.
+	outputPath := path.Join(c.RootPath, "config/crds")
+	if err := os.MkdirAll(outputPath, os.FileMode(0700)); err != nil {
+		return err
+	}
+
+	for file, crd := range crds {
+		generate.WriteString(path.Join(outputPath, file), crd)
+	}
+
+	return nil
+}
+
+func pathHasProjectFile(filePath string) bool {
+	if _, err := os.Stat(path.Join(filePath, "PROJECT")); os.IsNotExist(err) {
+		return false
+	}
+
+	return true
+}
+
+func isGoPath(filePath string) bool {
+	goPaths := strings.Split(os.Getenv("GOPATH"), ":")
+
+	for _, path := range goPaths {
+		if filePath == path {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getCRDFileName(resource *codegen.APIResource) string {
+	elems := []string{resource.Group, resource.Version, strings.ToLower(resource.Kind)}
+	return strings.Join(elems, "_") + ".yaml"
+}
+
+func (c *CRDGenerator) getCrds(p *parse.APIs) map[string]string {
+	crds := map[string]extensionsv1beta1.CustomResourceDefinition{}
+	for _, g := range p.APIs.Groups {
+		for _, v := range g.Versions {
+			for _, r := range v.Resources {
+				crd := r.CRD
+				if len(c.Namespace) > 0 {
+					crd.Namespace = c.Namespace
+				}
+				fileName := getCRDFileName(r)
+				crds[fileName] = crd
+			}
+		}
+	}
+
+	result := map[string]string{}
+	for file, crd := range crds {
+		s, err := yaml.Marshal(crd)
+		if err != nil {
+			glog.Fatalf("Error: %v", err)
+		}
+		result[file] = string(s)
+	}
+
+	return result
+}
+
+func getDomainFromProject() string {
+	var domain string
+
+	file, err := os.Open("./PROJECT")
+	if err != nil {
+		glog.Fatal(err)
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		if strings.HasPrefix(scanner.Text(), "domain:") {
+			domainInfo := strings.Split(scanner.Text(), ":")
+			if len(domainInfo) != 2 {
+				glog.Fatalf("Unexpected domain info: %s", scanner.Text())
+			}
+			domain = strings.Replace(domainInfo[1], " ", "", -1)
+			break
+		}
+	}
+
+	return domain
```

Comment:
We should be loading the file as `Go struct` to simplify this. I will check if we have something like that or not already.